File: go-algorithm/leetcode/problem5.go
package main

func longestPalindrome(s string) string {
	n := len(s)
	maxLen := 0
	ans := ""
	for i := 0; i < n; i++ {
		oddStr := expand(s, i, i)
		evenStr := expand(s, i, i+1)
		if len(oddStr) > maxLen {
			ans = oddStr
			maxLen = len(oddStr)
		}
		if len(evenStr) > maxLen {
			ans = evenStr
			maxLen = len(evenStr)
		}
	}
	return ans
}

func expand(s string, left int, right int) string {
	for left >= 0 && right < len(s) && s[left] == s[right] {
		left--
		right++
	}
	return s[left+1 : right]
}

File: java-algo/src/main/java/com/leetcode/Problem5.java
package com.leetcode;

/**
 * 5. Longest Palindromic Substring
 * https://leetcode.com/problems/longest-palindromic-substring/
 */
public class Problem5 {

    public String longestPalindrome(String s) {
        int len = s.length();
        if (len <= 1) {
            return s;
        }
        int maxLen = 0;
        String palindrome = "";

        for (int i = 0; i < len - 1; i++) {
            // odd
            String oddStr = check(s, i, i, len);
            int oddLen = oddStr.length();
            if (oddLen > maxLen) {
                maxLen = oddLen;
                palindrome = oddStr;
            }
            // even
            String evenStr = check(s, i, i + 1, len);
            int evenLen = evenStr.length();
            if (evenLen > maxLen) {
                maxLen = evenLen;
                palindrome = evenStr;
            }
        }

        return palindrome;
    }

    private String check(String s, int start, int end, int len) {
        while (start >= 0 && end <= len - 1 && s.charAt(start) == s.charAt(end)) {
            start--;
            end++;
        }
        return s.substring(start + 1, end);
    }

    public String longestPalindrome2(String s) {
        int maxLen = 0, n = s.length();
        String ans = "";
        for (int i = 0; i < n; i++) {
            String a = expand(s, i, i);
            String b = expand(s, i, i + 1);
            String temp = a.length() > b.length() ? a : b;
            if (temp.length() > maxLen) {
                maxLen = temp.length();
                ans = temp;
            }
        }
        return ans;
    }

    private String expand(String s, int left, int right) {
        while (left >= 0 && right < s.length() && s.charAt(left) == s.charAt(right)) {
            left--;
            right++;
        }
        return s.substring(left + 1, right);
    }
}
